File: risc0/circuit/rv32im/benches/eval_check.rs
// Copyright 2023 RISC Zero, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use risc0_circuit_rv32im::{
    cpu::CpuCircuitHal,
    testutil::{eval_check_impl, EvalCheckParams},
    CircuitImpl,
};
use risc0_core::field::baby_bear::BabyBear;
use risc0_zkp::{core::hash::sha::Sha256HashSuite, hal::cpu::CpuHal};

pub fn eval_check(c: &mut Criterion) {
    let mut group = c.benchmark_group("eval_check");
    group.sample_size(10);
    for po2 in [2, 8, 16].iter() {
        let params = EvalCheckParams::new(*po2);
        let circuit = CircuitImpl::new();
        let hal = CpuHal::new(Sha256HashSuite::<BabyBear>::new_suite());
        let circuit_hal = CpuCircuitHal::new(&circuit);
        group.bench_function(BenchmarkId::new("cpu", po2), |b| {
            b.iter(|| {
                eval_check_impl(&params, &hal, &circuit_hal);
            });
        });
    }

    #[cfg(feature = "cuda")]
    for po2 in [2, 8, 16, 20, 21].iter() {
        let params = EvalCheckParams::new(*po2);
        let hal = std::rc::Rc::new(risc0_zkp::hal::cuda::CudaHalSha256::new());
        let circuit_hal = risc0_circuit_rv32im::cuda::CudaCircuitHalSha256::new(hal.clone());
        group.bench_function(BenchmarkId::new("cuda", po2), |b| {
            b.iter(|| {
                eval_check_impl(&params, hal.as_ref(), &circuit_hal);
            });
        });
    }

    #[cfg(feature = "metal")]
    for po2 in [2, 8, 16, 22].iter() {
        let params = EvalCheckParams::new(*po2);
        let hal = std::rc::Rc::new(risc0_zkp::hal::metal::MetalHalSha256::new());
        let circuit_hal = risc0_circuit_rv32im::metal::MetalCircuitHal::<
            risc0_zkp::hal::metal::MetalHashSha256,
        >::new(hal.clone());
        group.bench_function(BenchmarkId::new("metal", po2), |b| {
            b.iter(|| {
                eval_check_impl(&params, hal.as_ref(), &circuit_hal);
            });
        });
    }
}

criterion_group!(benches, eval_check);
criterion_main!(benches);

File: risc0/circuit/rv32im-m3-sys/cxx/hal/cpu/eval_check.cpp
// Copyright 2025 RISC Zero, Inc.
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include "core/log.h"
#include "core/util.h"
#include "hal/hal.h"
#include "hal/po2s.h"
#include "zkp/fp.h"
#include "zkp/rou.h"

namespace risc0 {

#define PO2(x)                                                                                     \
  void eval_check_cpu_##x(Fp* check,                                                               \
                          const Fp* data,                                                          \
                          const Fp* accum,                                                         \
                          const Fp* globals,                                                       \
                          const FpExt* accMix,                                                     \
                          FpExt ecMix,                                                             \
                          Fp rou);
PO2S
#undef PO2

    void
    evalCheckCpu(Fp* check,
                 const Fp* data,
                 const Fp* accum,
                 const Fp* globals,
                 const FpExt* accMix,
                 FpExt ecMix,
                 size_t numRows) {
  // -2 do to compensating for expansion factor
  size_t po2 = log2Ceil(numRows) - 2;
  if (po2 < MIN_PO2 || po2 > MAX_PO2) {
    LOG(0, "PO2 = " << po2);
    throw std::runtime_error("Invalid po2");
  }
  if (numRows != (size_t(1) << (po2 + 2))) {
    throw std::runtime_error("numRows is not a power of 2");
  }

  switch (po2) {
#define PO2(x)                                                                                     \
  case x:                                                                                          \
    eval_check_cpu_##x(check, data, accum, globals, accMix, ecMix, ROU_FWD[po2 + 2]);              \
    break;
    PO2S
#undef PO2
        default : throw std::runtime_error("Unreachable");
  }
}

} // namespace risc0
